File: python/task3c/S3_PIPE_02.py
# S3_PIPE_02
from functools import reduce

def compose(*fns):
    # apply right to left
    return lambda x: reduce(lambda v, f: f(v), reversed(fns), x)

def inc(x): return x + 1
def double(x): return x * 2

# 5 * 2 + 1 = 11
pipeline = compose(inc, double)
print(pipeline(5))

File: js/task3/S3_PIPE_02.js
// S3_PIPE_02

const compose = (...fns) => x => fns.reduceRight((v, f) => f(v), x);

const inc = x => x + 1;
const double = x => x * 2;

// 5 * 2 + 1 = 11 (double first, then inc)
const func = compose(inc, double);
console.log(func(5));